libraries/AP_AIS/AP_AIS.cpp

```diff
@@ -0,0 +1,765 @@
+                    buffer_shift(msg_parts[i]);
+                }
+                if (!decoded && log_unsupported) {
+                    log_raw(&_incoming);
+                }
+            } else {
+                // multi part message, store in buffer
+                bool fits_in = false;
+                for  (uint8_t i = 0; i < AIVDM_BUFFER_SIZE; i++) {
+                    // find the first free spot
+                    if (_AIVDM_buffer[i].num == 0 && _AIVDM_buffer[i].total == 0 && _AIVDM_buffer[i].ID == 0) {
+                        _AIVDM_buffer[i] = _incoming;
+                        fits_in = true;
+                        break;
+                    }
+                }
+                if (!fits_in) {
+                    // remove the oldest message
+                    if (log_unsupported) {
+                        // log the unused message before removing it
+                        log_raw(&_AIVDM_buffer[0]);
+                    }
+                    buffer_shift(0);
+                    _AIVDM_buffer[AIVDM_BUFFER_SIZE - 1] = _incoming;
+                }
+            }
+        }
+    }
+
+    // remove expired items from the list
+    const uint32_t now =  AP_HAL::millis();
+    const uint32_t timeout = _time_out * 1000;
+    if (now < timeout) {
+        return;
+    }
+    const uint32_t deadline = now - timeout;
+    for (uint16_t i = 0; i < _list.max_items(); i++) {
+        if (_list[i].last_update_ms < deadline && _list[i].last_update_ms != 0) {
+            clear_list_item(i);
+        }
+    }
+}
+
+// Send a AIS mavlink message
+void AP_AIS::send(mavlink_channel_t chan)
+{
+    if (!enabled()) {
+        return;
+    }
+
+    const uint16_t list_size = _list.max_items();
+    const uint32_t now =  AP_HAL::millis();
+    uint16_t search_length = 0;
+    while (search_length < list_size) {
+        _send_index++;
+        search_length++;
+        if (_send_index == list_size) {
+            _send_index = 0;
+        }
+        if (_list[_send_index].last_update_ms != 0 &&
+            (_list[_send_index].last_send_ms < _list[_send_index].last_update_ms || now -_list[_send_index].last_send_ms > MAVLINK_RESEND_TIME_MS)) {
+                // only re-send if there has been a change or the resend time has expired
+                _list[_send_index].last_send_ms = now;
+                _list[_send_index].info.tslc = (now - _list[_send_index].last_update_ms) * 0.001;
+                mavlink_msg_ais_vessel_send_struct(chan,&_list[_send_index].info);
+                return;
+        }
+    }
+}
+
+// removed the given index from the AIVDM buffer shift following elements
+void AP_AIS::buffer_shift(uint8_t i)
+{
+    for (uint8_t n = i;  n < (AIVDM_BUFFER_SIZE - 1); n++) {
+        _AIVDM_buffer[n].ID = _AIVDM_buffer[n+1].ID;
+        _AIVDM_buffer[n].num = _AIVDM_buffer[n+1].num;
+        _AIVDM_buffer[n].total = _AIVDM_buffer[n+1].total;
+        strncpy(_AIVDM_buffer[n].payload,_AIVDM_buffer[n+1].payload,65);
```

Comment:
Is this actually safe?  Do you need `memmove`?

How sure are you over buffer termination here, because 65 is kinda interesting, too.